### ISSUE TITLE
Can't pickle function delayed

### ISSUE THREAD

**Issue by vincentschut** [dask/distributed#725]
I get a "_pickle.PicklingError: Can't pickle <function delayed at 0x7fceb5c89488>" when trying to use calculate on an instance method that returns a delayed. Afair this used to work until some days ago...

Minimal example:

```
from dask import delayed
from distributed import Client

class TestClass():
    def __init__(self):
        pass

    def do_something(self, value):
        return value + 10

    def get_delayed(self, value):
        return delayed(self.do_something)(value)


if __name__ == '__main__':
    c = Client('127.0.0.1:8786') # I have a scheduler and workers running locally

    tc = TestClass()

    print(c)

    d = tc.get_delayed(10)
    r = d.compute()

    print(r)
```

I'm using dask and distrubuted from git master:
In [3]: print(dask.__version__)
0.12.0+43.g7a2d5e3

In [4]: print(distributed.__version__)
1.14.3+77.gaba8ac2

This is the traceback I get:

```
<Client: scheduler="127.0.0.1:8786" processes=1 cores=8>
Traceback (most recent call last):
  File "/usr/lib/python3.5/site-packages/distributed/protocol/pickle.py", line 38, in dumps
    return cloudpickle.dumps(x, protocol=pickle.HIGHEST_PROTOCOL)
  File "/usr/lib/python3.5/site-packages/cloudpickle/cloudpickle.py", line 629, in dumps
    cp.dump(obj)
  File "/usr/lib/python3.5/site-packages/cloudpickle/cloudpickle.py", line 107, in dump
    return Pickler.dump(self, obj)
  File "/usr/lib64/python3.5/pickle.py", line 408, in dump
    self.save(obj)
  File "/usr/lib64/python3.5/pickle.py", line 475, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib/python3.5/site-packages/cloudpickle/cloudpickle.py", line 381, in save_instancemethod
    self.save_reduce(types.MethodType, (obj.__func__, obj.__self__), obj=obj)
  File "/usr/lib/python3.5/site-packages/cloudpickle/cloudpickle.py", line 529, in save_reduce
    save(args)
  File "/usr/lib64/python3.5/pickle.py", line 475, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib64/python3.5/pickle.py", line 725, in save_tuple
    save(element)
  File "/usr/lib64/python3.5/pickle.py", line 520, in save
    self.save_reduce(obj=obj, *rv)
  File "/usr/lib/python3.5/site-packages/cloudpickle/cloudpickle.py", line 514, in save_reduce
    save(cls)
  File "/usr/lib64/python3.5/pickle.py", line 475, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib/python3.5/site-packages/cloudpickle/cloudpickle.py", line 368, in save_global
    self.save_reduce(typ, (obj.__name__, obj.__bases__, d), obj=obj)
  File "/usr/lib/python3.5/site-packages/cloudpickle/cloudpickle.py", line 529, in save_reduce
    save(args)
  File "/usr/lib64/python3.5/pickle.py", line 475, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib64/python3.5/pickle.py", line 725, in save_tuple
    save(element)
  File "/usr/lib64/python3.5/pickle.py", line 475, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib64/python3.5/pickle.py", line 810, in save_dict
    self._batch_setitems(obj.items())
  File "/usr/lib64/python3.5/pickle.py", line 836, in _batch_setitems
    save(v)
  File "/usr/lib64/python3.5/pickle.py", line 475, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib/python3.5/site-packages/cloudpickle/cloudpickle.py", line 205, in save_function
    self.save_function_tuple(obj)
  File "/usr/lib/python3.5/site-packages/cloudpickle/cloudpickle.py", line 253, in save_function_tuple
    save(f_globals)
  File "/usr/lib64/python3.5/pickle.py", line 475, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib64/python3.5/pickle.py", line 810, in save_dict
    self._batch_setitems(obj.items())
  File "/usr/lib64/python3.5/pickle.py", line 841, in _batch_setitems
    save(v)
  File "/usr/lib64/python3.5/pickle.py", line 506, in save
    self.save_global(obj, rv)
  File "/usr/lib/python3.5/site-packages/cloudpickle/cloudpickle.py", line 370, in save_global
    raise pickle.PicklingError("Can't pickle %r" % obj)
_pickle.PicklingError: Can't pickle <function delayed at 0x7fceb5c89488>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "trial_delayed_001.py", line 28, in <module>
    r = d.compute()
  File "/usr/lib/python3.5/site-packages/dask/base.py", line 79, in compute
    return compute(self, **kwargs)[0]
  File "/usr/lib/python3.5/site-packages/dask/base.py", line 179, in compute
    results = get(dsk, keys, **kwargs)
  File "/usr/lib/python3.5/site-packages/distributed/client.py", line 1358, in get
    restrictions, loose_restrictions, resources=resources)
  File "/usr/lib/python3.5/site-packages/distributed/client.py", line 1301, in _graph_to_futures
    'tasks': valmap(dumps_task, dsk3),
  File "/usr/lib/python3.5/site-packages/toolz/dicttoolz.py", line 84, in valmap
    rv.update(zip(iterkeys(d), map(func, itervalues(d))))
  File "/usr/lib/python3.5/site-packages/distributed/worker.py", line 608, in dumps_task
    return {'function': dumps_function(task[0]),
  File "/usr/lib/python3.5/site-packages/distributed/worker.py", line 575, in dumps_function
    b = dumps(func)
  File "/usr/lib/python3.5/site-packages/distributed/protocol/pickle.py", line 48, in dumps
    return cloudpickle.dumps(x, protocol=pickle.HIGHEST_PROTOCOL)
  File "/usr/lib/python3.5/site-packages/cloudpickle/cloudpickle.py", line 629, in dumps
    cp.dump(obj)
  File "/usr/lib/python3.5/site-packages/cloudpickle/cloudpickle.py", line 107, in dump
    return Pickler.dump(self, obj)
  File "/usr/lib64/python3.5/pickle.py", line 408, in dump
    self.save(obj)
  File "/usr/lib64/python3.5/pickle.py", line 475, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib/python3.5/site-packages/cloudpickle/cloudpickle.py", line 381, in save_instancemethod
    self.save_reduce(types.MethodType, (obj.__func__, obj.__self__), obj=obj)
  File "/usr/lib/python3.5/site-packages/cloudpickle/cloudpickle.py", line 529, in save_reduce
    save(args)
  File "/usr/lib64/python3.5/pickle.py", line 475, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib64/python3.5/pickle.py", line 725, in save_tuple
    save(element)
  File "/usr/lib64/python3.5/pickle.py", line 520, in save
    self.save_reduce(obj=obj, *rv)
  File "/usr/lib/python3.5/site-packages/cloudpickle/cloudpickle.py", line 514, in save_reduce
    save(cls)
  File "/usr/lib64/python3.5/pickle.py", line 475, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib/python3.5/site-packages/cloudpickle/cloudpickle.py", line 368, in save_global
    self.save_reduce(typ, (obj.__name__, obj.__bases__, d), obj=obj)
  File "/usr/lib/python3.5/site-packages/cloudpickle/cloudpickle.py", line 529, in save_reduce
    save(args)
  File "/usr/lib64/python3.5/pickle.py", line 475, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib64/python3.5/pickle.py", line 725, in save_tuple
    save(element)
  File "/usr/lib64/python3.5/pickle.py", line 475, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib64/python3.5/pickle.py", line 810, in save_dict
    self._batch_setitems(obj.items())
  File "/usr/lib64/python3.5/pickle.py", line 836, in _batch_setitems
    save(v)
  File "/usr/lib64/python3.5/pickle.py", line 475, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib/python3.5/site-packages/cloudpickle/cloudpickle.py", line 205, in save_function
    self.save_function_tuple(obj)
  File "/usr/lib/python3.5/site-packages/cloudpickle/cloudpickle.py", line 253, in save_function_tuple
    save(f_globals)
  File "/usr/lib64/python3.5/pickle.py", line 475, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib64/python3.5/pickle.py", line 810, in save_dict
    self._batch_setitems(obj.items())
  File "/usr/lib64/python3.5/pickle.py", line 841, in _batch_setitems
    save(v)
  File "/usr/lib64/python3.5/pickle.py", line 506, in save
    self.save_global(obj, rv)
  File "/usr/lib/python3.5/site-packages/cloudpickle/cloudpickle.py", line 370, in save_global
    raise pickle.PicklingError("Can't pickle %r" % obj)
_pickle.PicklingError: Can't pickle <function delayed at 0x7fceb5c89488>
```

**Comment by mrocklin** [comment]
Thank you @vincentschut for the excellent reproducible example.  This is really helpful.  I've shrunk it down to the following case:

```python
In [1]: from dask import delayed
   ...: 
   ...: class TestClass():
   ...:     def __init__(self):
   ...:         pass
   ...: 
   ...:     def do_something(self, value):
   ...:         return value + 10
   ...: 
   ...:     def get_delayed(self, value):
   ...:         return delayed(self.do_something)(value)
   ...:     

In [2]: from cloudpickle import dumps, loads

In [3]: tc = TestClass()

In [4]: loads(dumps(tc.do_something))
Out[4]: <bound method TestClass.do_something of <__main__.TestClass object at 0x7f5f339a7470>>

In [5]: from dask import delayed
   ...: from distributed import Client
   ...: 
   ...: class TestClass():
   ...:     def __init__(self):
   ...:         pass
   ...: 
   ...:     def do_something(self, value):
   ...:         return value + 10
   ...: 
   ...:     def get_delayed(self, value):
   ...:         return delayed(self.do_something)(value)
   ...:     

In [6]: tc = TestClass()

In [7]: loads(dumps(tc.do_something))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-7-47ea49f231fa> in <module>()
----> 1 loads(dumps(tc.do_something))

TypeError: Required argument 'mapping' (pos 1) not found
```

So everything works fine until we import `distributed`

When we dump just `delayed` we get a long bytestring, implying that cloudpickle isn't able to serialize it trivially

```
In [11]: dumps(delayed)
Out[11]: b'\x80\x02ctoolz.functoolz\ncurry\nq\x00)\x81q\x01(ccloudpickle.cloudpickle\n_fill_function\nq\x02(ccloudpickle.cloudpickle\n_make_skel_func\nq\x03ccloudpickle.cloudpickle\n_builtin_type\nq\x04X\x08\x00\x00\x00CodeTypeq\x05\x85q\x06Rq\x07(K\x04K\x00K\x08K\x0fKCc_codecs\nencode\nq\x08XB\x01\x00\x00t\x00\x00|\x00\x00t\x01\x00\xc2\x83\x02\x00r\x13\x00|\x00\x00St\x02\x00|\x00\x00\xc2\x83\x01\x00\\\x02\x00}\x04\x00}\x05\x00|\x05\x00s\xc3\xa0\x00|\x03\x00d\x01\x00k\x08\x00pR\x00t\x03\x00|\x03\x00\xc2\x83\x01\x00t\x04\x00k\x08\x00oR\x00|\x03\x00d\x02\x00k\x05\x00sh\x00t\x05\x00d\x03\x00|\x03\x00\x16\xc2\x83\x01\x00\xc2\x82\x01\x00n\x00\x00|\x01\x00s\xc3\x87\x00y\r\x00|\x00\x00j\x06\x00}\x06\x00Wn!\x00\x04t\x07\x00k\n\x00r\xc2\x9e\x00\x01\x01\x01t\x03\x00|\x00\x00\xc2\x83\x01\x00j\x06\x00}\x06\x00Yn\x01\x00Xt\x08\x00|\x00\x00|\x03\x00d\x04\x00|\x02\x00\xc2\x83\x02\x01}\x07\x00d\x05\x00|\x06\x00|\x07\x00f\x02\x00\x16}\x01\x00n\x00\x00t\t\x00|\x00\x00|\x01\x00d\x04\x00|\x02\x00d\x06\x00|\x03\x00\xc2\x83\x02\x02S|\x01\x00s\x0e\x01d\x05\x00t\x03\x00|\x00\x00\xc2\x83\x01\x00j\x06\x00t\x08\x00|\x04\x00d\x04\x00|\x02\x00\xc2\x83\x01\x01f\x02\x00\x16}\x01\x00n\x00\x00|\x05\x00j\n\x00i\x01\x00|\x04\x00|\x01\x006\xc2\x83\x01\x00\x01t\x01\x00|\x01\x00|\x05\x00\xc2\x83\x02\x00Sd\x01\x00Sq\tX\x06\x00\x00\x00latin1q\n\x86q\x0bRq\x0c(X\xb0\x10\x00\x00Wraps a function or object to produce a ``Delayed``.\n\n    ``Delayed`` objects act as proxies for the object they wrap, but all\n    operations on them are done lazily by building up a dask graph internally.\n\n    Parameters\n    ----------\n    obj : object\n        The function or object to wrap\n    name : string or hashable, optional\n        The key to use in the underlying graph for the wrapped object. Defaults\n        to hashing content.\n    pure : bool, optional\n        Indicates whether calling the resulting ``Delayed`` object is a pure\n        operation. If True, arguments to the call are hashed to produce\n        deterministic keys. Default is False.\n    nout : int, optional\n        The number of outputs returned from calling the resulting ``Delayed``\n        object. If provided, the ``Delayed`` output of the call can be iterated\n        into ``nout`` objects, allowing for unpacking of results. By default\n        iteration over ``Delayed`` objects will error.\n\n    Examples\n    --------\n    Apply to functions to delay execution:\n\n    >>> def inc(x):\n    ...     return x + 1\n\n    >>> inc(10)\n    11\n\n    >>> x = delayed(inc, pure=True)(10)\n    >>> type(x) == Delayed\n    True\n    >>> x.compute()\n    11\n\n    Can be used as a decorator:\n\n    >>> @delayed(pure=True)\n    ... def add(a, b):\n    ...     return a + b\n    >>> add(1, 2).compute()\n    3\n\n    ``delayed`` also accepts an optional keyword ``pure``. If False (default),\n    then subsequent calls will always produce a different ``Delayed``. This is\n    useful for non-pure functions (such as ``time`` or ``random``).\n\n    >>> from random import random\n    >>> out1 = delayed(random, pure=False)()\n    >>> out2 = delayed(random, pure=False)()\n    >>> out1.key == out2.key\n    False\n\n    If you know a function is pure (output only depends on the input, with no\n    global state), then you can set ``pure=True``. This will attempt to apply a\n    consistent name to the output, but will fallback on the same behavior of\n    ``pure=False`` if this fails.\n\n    >>> @delayed(pure=True)\n    ... def add(a, b):\n    ...     return a + b\n    >>> out1 = add(1, 2)\n    >>> out2 = add(1, 2)\n    >>> out1.key == out2.key\n    True\n\n    The key name of the result of calling a delayed object is determined by\n    hashing the arguments by default. To explicitly set the name, you can use\n    the ``dask_key_name`` keyword when calling the function:\n\n    >>> add(1, 2)    # doctest: +SKIP\n    Delayed(\'add-3dce7c56edd1ac2614add714086e950f\')\n    >>> add(1, 2, dask_key_name=\'three\')\n    Delayed(\'three\')\n\n    Note that objects with the same key name are assumed to have the same\n    result. If you set the names explicitly you should make sure your key names\n    are different for different results.\n\n    >>> add(1, 2, dask_key_name=\'three\')\n    >>> add(2, 1, dask_key_name=\'three\')\n    >>> add(2, 2, dask_key_name=\'four\')\n\n    ``delayed`` can also be applied to objects to make operations on them lazy:\n\n    >>> a = delayed([1, 2, 3])\n    >>> isinstance(a, Delayed)\n    True\n    >>> a.compute()\n    [1, 2, 3]\n\n    The key name of a delayed object is hashed by default if ``pure=True`` or\n    is generated randomly if ``pure=False`` (default).  To explicitly set the\n    name, you can use the ``name`` keyword:\n\n    >>> a = delayed([1, 2, 3], name=\'mylist\')\n    >>> a\n    Delayed(\'mylist\')\n\n    Delayed results act as a proxy to the underlying object. Many operators\n    are supported:\n\n    >>> (a + [1, 2]).compute()\n    [1, 2, 3, 1, 2]\n    >>> a[1].compute()\n    2\n\n    Method and attribute access also works:\n\n    >>> a.count(2).compute()\n    1\n\n    Note that if a method doesn\'t exist, no error will be thrown until runtime:\n\n    >>> res = a.not_a_real_method()\n    >>> res.compute()  # doctest: +SKIP\n    AttributeError("\'list\' object has no attribute \'not_a_real_method\'")\n\n    Methods are assumed to be impure by default, meaning that subsequent calls\n    may return different results. To assume purity, set `pure=True`. This\n    allows sharing of any intermediate values.\n\n    >>> a.count(2, pure=True).key == a.count(2, pure=True).key\n    True\n\n    As with function calls, method calls also support the ``dask_key_name``\n    keyword:\n\n    >>> a.count(2, dask_key_name="count_2")\n    Delayed("count_2")\n\n    q\rNK\x00X/\x00\x00\x00nout must be None or a positive integer, got %sq\x0eX\x04\x00\x00\x00pureq\x0fX\x05\x00\x00\x00%s-%sq\x10X\x04\x00\x00\x00noutq\x11tq\x12(X\n\x00\x00\x00isinstanceq\x13X\x07\x00\x00\x00Delayedq\x14X\r\x00\x00\x00to_task_dasksq\x15X\x04\x00\x00\x00typeq\x16X\x03\x00\x00\x00intq\x17X\n\x00\x00\x00ValueErrorq\x18X\x08\x00\x00\x00__name__q\x19X\x0e\x00\x00\x00AttributeErrorq\x1aX\x08\x00\x00\x00tokenizeq\x1bX\x0b\x00\x00\x00DelayedLeafq\x1cX\x06\x00\x00\x00appendq\x1dtq\x1e(X\x03\x00\x00\x00objq\x1fX\x04\x00\x00\x00nameq h\x0fh\x11X\x04\x00\x00\x00taskq!X\x05\x00\x00\x00dasksq"X\x06\x00\x00\x00prefixq#X\x05\x00\x00\x00tokenq$tq%X-\x00\x00\x00/home/mrocklin/workspace/dask/dask/delayed.pyq&X\x07\x00\x00\x00delayedq\'Kmh\x08X\'\x00\x00\x00\x00\xc2\x89\x0f\x01\x04\x02\x12\x02\x06\x01*\x01\x06\x01\r\x01\x06\x01\x03\x01\r\x01\r\x01\x14\x01\x15\x01\x13\x01\x19\x02\x06\x01(\x01\x14\x01q(h\n\x86q)Rq*))tq+Rq,]q-}q.\x87q/Rq0}q1(h\x1bcdask.delayed\ntokenize\nq2h\x15cdask.delayed\nto_task_dasks\nq3h\x1ccdask.delayed\nDelayedLeaf\nq4h\x14cdask.delayed\nDelayed\nq5uN\x89N\x87q6}q7tR)}q8(X\x07\x00\x00\x00__doc__q9h\r\x86q:X\x08\x00\x00\x00_sigspecq;cinspect\nSignature\nq<)\x81q=N}q>(X\x0b\x00\x00\x00_parametersq?h\x04X\x10\x00\x00\x00MappingProxyTypeq@\x85qARqB)\x81qCX\x12\x00\x00\x00_return_annotationqDcinspect\n_empty\nqEu\x86qFb\x86qGX\x11\x00\x00\x00_has_unknown_argsqH\x89\x86qIh\x19h\'\x86qJtqKtqLb.'
```

So it appears that serializing classes that contain delayed is causing issues.  Presumably you could resolve this short term by importing delayed from within the method:


```python
In [12]: 
    ...: from distributed import Client
    ...: 
    ...: class TestClass():
    ...:     def __init__(self):
    ...:         pass
    ...: 
    ...:     def do_something(self, value):
    ...:         return value + 10
    ...: 
    ...:     def get_delayed(self, value):
    ...:         from dask import delayed
    ...:         return delayed(self.do_something)(value)
    ...:     

In [13]: tc = TestClass()

In [14]: loads(dumps(tc.do_something))
Out[14]: <bound method TestClass.do_something of <__main__.TestClass object at 0x7fb8bebd8c50>>
```

But long term we should do a few things:

1.  Figure out what is happening at `distributed` import time that is making serialization behaviors change.
2.  Perhaps stop using `toolz.curry` on `dask.delayed` and instead curry manually by changing the signature to `def delayed(obj='-no-default-', pure=None, ...)` so that the function itself becomes simpler to serialize.

*If* you wanted to dive into this issue, the next thing I would do is to bisect the `distributed` commit history and try to find the point at which this change occurs.  That would probably help us to identify the original cause of the change.

**Comment by vincentschut** [comment]
Thanks. The workaround _seems_ to work, but I'm running into other issues now. I'll try to separate those out and file issues for each of them. I _might_ have time to do some bisecting later, but please don't count on it.
Keeping this open for now.

**Comment by pitrou** [comment]
It appears `curry` isn't propagating function metadata correctly, unless it's deliberate:
```
>>> import pickle, cloudpickle
>>> from dask import delayed
>>> delayed.__module__
'toolz.functoolz'
>>> delayed.__name__
'delayed'
```

Note the pickle error message is a bit misleading:
```
>>> pickle.dumps(delayed)
Traceback (most recent call last):
  File "<ipython-input-5-2e57bc4a5b42>", line 1, in <module>
    pickle.dumps(delayed)
PicklingError: Can't pickle <function delayed at 0x7f15945e9b70>: it's not the same object as dask.delayed.delayed

>>> import sys
>>> sys.modules['dask.delayed'].delayed
<function delayed at 0x7f15945e9b70>
>>> sys.modules['dask.delayed'].delayed is delayed
True
```

**Comment by mrocklin** [comment]
cc @eriknw

**Comment by eriknw** [comment]
@pitrou nailed this on the head.  We need to set `__module__` in `toolz.curry`, then it works as expected:
```python
>>> import pickle
>>> from dask import delayed
>>> delayed.__module__
'dask.delayed'
>>> delayed2 = pickle.loads(pickle.dumps(delayed))
>>> delayed is delayed2
True
```

**Comment by eriknw** [comment]
pytoolz/toolz#355 should fix this.  Is this worthy of a bug-fix release?

**Comment by eriknw** [comment]
This should be fixed by installing `toolz=0.8.2`.
